File: LeetCode_Python/BackTracking/__Permutation/Leetcode_51_n_queens/Leetcode_51_n_queens.py
"""
@brief: Leetcode_51_N_皇后
@link: https://leetcode.cn/problems/n-queens/description/
@author: liangj.zhang
@date: 14/3/2026

@Difficulty: Hard

@Label: Backtracking & permutation

@Retrospect(worthy 1 - 5): 5

@thoughts:
    + 【思路 1】：回溯
        是在按照灵神的题单在刷，这里是视频讲解题，但有想法的，我习惯在视频之前按照自己的思路来一遍（如果有的话）
        ----
        这题自己的思路，就是题目，这个游戏的规则，结合回溯的枚举特性，
        每个皇后不会在同一横向和纵向，以及斜线；
        我就枚举每一行位置的可能，
            枚举第一行位置的可能，
            下一行从没有选过的列当中再选一个，下下一行亦是如此；
        ----
        最先完成的一版有重复的问题，再结合 chat 的分析之后，改了过来；
        原来的行号也是通过 for 循环一次选出一个（所以第一个行号就有 4 中情况...），没有按固定顺序枚举下来；有这么个问题；

        + 分析：
            + 时间复杂度：O(n! * n)，n! -> 是情况的可能性，n 是每个节点都要遍历 n 次
            + 空间复杂度：不算存储结果：O(n)
        + rank:
            + 时间效率：15 ms, 击败 42.65%
            + 空间效率：19.42 MB, 击败 53.91%
"""
from typing import List


class Solution:
    
    def solveNQueens(self, n: int) -> List[List[str]]:
        col_s = set()

        path = []
        ans = []
        def dfs(idx: int) -> None:
            nonlocal col_s, path, ans
            if idx == n:
                tmp = []
                for i in range(len(path)):
                    row = ['.'] * n
                    row[path[i][1]] = 'Q'
                    tmp.append(''.join(row))
                ans.append(tmp.copy())
                return

            x = idx # 原来错误的写法：给 x 也来了一遍 for 循环
            for y in range(n):
                if y not in col_s:
                    find = True
                    for x_, y_ in path:
                        if abs(y_ - y) == abs(x_ - x):
                            find = False
                            break
                    if find:
                        col_s.add(y)
                        path.append((x,y))

                        dfs(idx + 1)

                        path.pop()
                        col_s.remove(y)
        dfs(0)
        return ans
    
print(Solution().solveNQueens(n=1))

File: LeetCode/BackTracking/__Permutation/Leetcode_51_n_queens/Leetcode_51_n_queens.cpp
/**
 * @brief: Leetcode_51_N_皇后
 * @link: https://leetcode.cn/problems/n-queens/description/
 * @author: liangj.zhang
 * @date: 14/3/2026
 * 
 * @updated:
 *  + 28/3/2026: add
 * 
 * @Difficulty: Hard
 * 
 * @Label: Backtracking
 * 
 * @Retrospect(worthy 1 - 5): 5
 * 
 * @thoughts: 
 *  + 【思路 2】：回溯
 *      python 写得第一版，是通过存储每一行坐标，然后跟当前行的坐标相减，求斜率的绝对值，如果是 1，则说明这个位置不能选；
 *      列都是通过集合来排除；
 * 
 *      这里斜线的比较是通过，坐标特性：
 *          左斜线：y - x 是恒定的，同一斜线上；
 *          有些先：y + x 是恒定的；
 *      + 分析：
 *          + 时间复杂度：O(n! * n) -> 后面的 n 是copy的那份，当然拷贝的次数，肯定是远低于 n!
 *          + 空间复杂度：递归 O(n)， 结果拷贝：O(n^2), 辅助结构 O(n)，结果集 O(k * n^2), k 是 结果个数
 *      + rank:
 *          + 时间效率：12 ms, 击败 18.36%
 *          + 空间效率：13.86 MB, 击败 12.37%
 *  + 【思路 1】：回溯
 *      同 python 写的那一版，存储坐标，然后
 *          1. 是否在同一列
 *          2. 是否在同一斜线：abs(path[i] - col) == abs(n - i)
 *      + 分析: 完全同上
 *      + rank:
 *          + 时间效率：3 ms, 击败 81.84%
 *          + 空间效率：12.39 MB, 击败 20.78%
 */

#include <vector>
#include <string>
using std::vector;
using std::string;

#include <unordered_set>
using std::unordered_set;
#define us unordered_set

class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        us<int> cols;
        us<int> leftUp;
        us<int> rightUp;

        vector<int> path;
        vector<vector<string>> ans;
        auto dfs = [&](this auto&& dfs, int row) -> void {

            if (row == n) {
                vector<string> tmp;
                for (int i = 0; i < n; ++i) {
                    vector<char> r(n, '.');
                    r[path[i]] = 'Q';
                    tmp.emplace_back(string(r.begin(), r.end()));
                }
                ans.emplace_back(tmp);
                return;
            }

            for (int col = 0; col < n; ++col) {
                // col 没选过 && 左斜不会冲突 && 右斜不会冲突
                if (int _sub = col - row, _sum = col + row;
                    !cols.contains(col) && !leftUp.contains(_sub) && !rightUp.contains(_sum)) {
                    cols.insert(col);
                    leftUp.insert(_sub);
                    rightUp.insert(_sum);
                    path.push_back(col);

                    dfs(row + 1);

                    path.pop_back();
                    rightUp.erase(_sum);
                    leftUp.erase(_sub);
                    cols.erase(col);
                }
            }
        };

        dfs(0);

        return ans;
    }
};

class Solution {
private:
    vector<int> path;
    vector<vector<string>> ans;
    bool isConflict(int col) {
        int n = path.size();
        for (int i = 0; i < n; ++i) {
            // (i, path[i]) - (n, col)
            if (path[i] == col || abs(path[i] - col) == abs(n - i))
                return true;
        }
        return false;
    }
public:
    vector<vector<string>> solveNQueens(int n) {
        path.clear();
        ans.clear();

        auto dfs = [&](this auto&& dfs, int row) -> void {

            if (row == n) {
                vector<string> tmp;
                for (int i = 0; i < n; ++i) {
                    vector<char> r(n, '.');
                    r[path[i]] = 'Q';
                    tmp.emplace_back(std::move(string(r.begin(), r.end())));
                }
                ans.emplace_back(std::move(tmp));
                return;
            }

            for (int col = 0; col < n; ++col) {
                // col 没选过 && 左斜不会冲突 && 右斜不会冲突
                if (!isConflict(col)) {
                    path.push_back(col);

                    dfs(row + 1);

                    path.pop_back();
                }
            }
        };

        dfs(0);

        return ans;
    }
};